rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,416 @@
+    // If `vsock_id` was present, set the deprecation message in `parsing_info`.
+    if let Some(msg) = deprecation_message {
+        parsed_req.parsing_info().append_deprecation_message(msg);
+    }
+
+    Ok(parsed_req)
+}
+```
+
+In our proof, we commented out one line in the function that updates statistics (and that seems to lead to bad solver performance), and manually mocked some of the Firecracker types with simpler versions to reduce the number of dependencies we had to pull in (e.g., we removed some enum variants, unused struct fields).
+With these changes, we were able to prove that the configuration data has a vsock ID if and only if the parsing metadata includes a deprecation message: 
+
+```rust
+#[cfg(kani)]
+fn get_vsock_device_config(action: RequestAction) -> Option<VsockDeviceConfig> {
+    match action {
+        RequestAction::Sync(vmm_action) => match *vmm_action {
+            VmmAction::SetVsockDevice(dev) => Some(dev),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::unwind(5)]
+#[kani::stub_py("serde_json::deserialize_slice", "mock_deserialize")]
+fn test_deprecation_vsock_id_consistent() {
+    // We are going to mock the parsing of this body, so might as well use an empty one.
+    let body: Vec<u8> = Vec::new();
+    if let Ok(res) = parse_put_vsock(&Body::new(body)) {
+        let (action, mut parsing_info) = res.into_parts();
+        let config = get_vsock_device_config(action).unwrap();
+        assert_eq!(
+            config.vsock_id.is_some(),
+            parsing_info.take_deprecation_message().is_some()
+        );
+    }
+}
+```
+
+Crucially, we did this by stubbing out `serde_json::from_slice` and replacing it with our mock version below, which ignores its input and creates a "symbolic" configuration struct:
+
+```rust
+#[cfg(kani)]
+fn symbolic_string(len: usize) -> String {
+    let mut v: Vec<u8> = Vec::with_capacity(len);
+    for _ in 0..len {
+        v.push(kani::any());
+    }
+    unsafe { String::from_utf8_unchecked(v) }
+}
+
+#[cfg(kani)]
+fn mock_deserialize(_data: &[u8]) -> serde_json::Result<VsockDeviceConfig> {
+    let str_len = 4;
+    let vsock_id = if kani::any() {
+        None
+    } else {
+        Some(symbolic_string(str_len))
+    };
+    let guest_cid = kani::any();
+    let uds_path = symbolic_string(str_len);
+    let dev = VsockDeviceConfig {
+        vsock_id,
+        guest_cid,
+        uds_path,
+    };
+    Ok(dev)
+}
+```
+
+The proof takes 226 seconds to complete.
+
+## User Experience
+
+This feature is currently limited to stubbing functions and methods.
+We anticipate that the annotations we propose here could also be used when stubbing types, although the underlying technical approach might have to change.
+
+Stubs will be specified per harness; that is, different harnesses can use different stubs (the reasoning being that users might want to mock different behavior for different harnesses).
```

Comment:
Or maybe they don't want to mock its behavior. In general, you don't want to force the same stubs for a complete code base. Needs more explaining... this is the main design point.